File: Go/Go_100-999/761.go
// Go
// leetcode 761
// https://leetcode.cn/problems/special-binary-string/

func makeLargestSpecial(s string) string {
    candidates := sort.StringSlice{}
    for i, cur, last := 0, 0, 0; i < len(s); i++ {
        if s[i] == '1' {
            cur++
        } else {
            cur--
        }
        if cur == 0 {
            candidates = append(candidates, "1" + makeLargestSpecial(s[last + 1 : i]) + "0")
            last = i + 1
        }
    }
    sort.Sort(sort.Reverse(candidates))
    return strings.Join(candidates, "")
}

File: cpp/cpp_100-999/761.cc
// C++
// leetcode 761
// https://leetcode.cn/problems/special-binary-string/

class Solution {
public:
    string makeLargestSpecial(string s) {
    	int n = s.size();
    	vector<string> tmp;
    	for (int i = 0, cur = 0, last = 0; i < n; i++) {
    		if (s[i] == '1')
    			cur++;
    		else
    			cur--;
    		if (cur == 0) {
    			tmp.push_back("1" + makeLargestSpecial(s.substr(last + 1, i - last - 1)) + "0");
    			last = i + 1;
    		}
    	}
    	sort(tmp.begin(), tmp.end());
    	string res = "";
    	for(int i = tmp.size() - 1; i >= 0; i--)
    		res += tmp[i];
    	return res;
    }
};